[Transform] TransformIndexer change detector causes log spam and isn't integrated into retry logic

When transform isn't able to detect changes it logs and audits the following message as warning:

`[${transform_id}] failed to detect changes for transform. Skipping update till next check.`

On a broken cluster this causes a flood of log messages as this gets logged for every failure. The failure happens based on `frequency`. Apart from the log spam this check isn't properly integrated into retry and failure handlers. The later would prevent the spam. The failure handler lets the transform retry based on exponential back-off. A transform that isn't executed as `unattended` should fail after the configured number of retries (if the number of retries is configured as -1 it should retry forever).

The code in question can be found in `sourceHasChanged` (called by `onStart`) in `TransformIndexer`

Todo:

 - [ ] create a failure test case that simulates the error (see details below)
 - [ ] don't suppress the failure in `sourceHasChanged`, but let the error pass and be handled in `TransformFailureHandler`

Details:

The failure:

```
Caused by: org.elasticsearch.action.search.SearchPhaseExecutionException: Search rejected due to missing shards [[...][0], [...][0]]. Consider using `allow_partial_search_results` setting to bypass this error.
    at org.elasticsearch.action.search.AbstractSearchAsyncAction.run(AbstractSearchAsyncAction.java:245) ~[elasticsearch-8.5.1.jar:?]
```

--- a/docs/changelog/92762.yaml
+++ b/docs/changelog/92762.yaml
@@ -1,0 +1,6 @@
+pr: 92762
+summary: Integrate "sourceHasChanged" call into failure handling and retry logic
+area: Transform
+type: bug
+issues:
+ - 92133

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
@@ -357,7 +357,7 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
         // If we are not on the initial batch checkpoint and its the first pass of whatever continuous checkpoint we are on,
         // we should verify if there are local changes based on the sync config. If not, do not proceed further and exit.
         if (context.getCheckpoint() > 0 && initialRun()) {
-            sourceHasChanged(ActionListener.wrap(hasChanged -> {
+            checkpointProvider.sourceHasChanged(getLastCheckpoint(), ActionListener.wrap(hasChanged -> {
                 context.setLastSearchTime(instantOfTrigger);
                 hasSourceChanged = hasChanged;
                 if (hasChanged) {
@@ -957,20 +957,6 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
             logger.debug("[{}] checked for outdated checkpoints", getJobId());
             listener.onResponse(null);
         }
-    }
-
-    private void sourceHasChanged(ActionListener<Boolean> hasChangedListener) {
-        checkpointProvider.sourceHasChanged(getLastCheckpoint(), ActionListener.wrap(hasChanged -> {
-            logger.trace("[{}] change detected [{}].", getJobId(), hasChanged);
-            hasChangedListener.onResponse(hasChanged);
-        }, e -> {
-            logger.warn(() -> "[" + getJobId() + "] failed to detect changes for transform. Skipping update till next check.", e);
-            auditor.warning(
-                getJobId(),
-                "Failed to detect changes for transform, skipping update till next check. Exception: " + e.getMessage()
-            );
-            hasChangedListener.onResponse(false);
-        }));
     }
 
     private IterationResult<TransformIndexerPosition> processBuckets(final SearchResponse searchResponse) {

